### DISCUSSION TITLE
Multi-Tenacy With DataSources, OpenShift

### DISCUSSION THREAD

**Discussion by Lucostus** [grafana/grafana#58409] [Platform]
With the license change of Elasticsearch, OpenShift switches to the Loki stack, Grafana is not supported by Red Hat itself, but Grafana is the only decent solution to visualize the logs + metrics. 

OpenShift puts an Auth Proxy in front of Thanos and Loki (this ensures that normal users can only monitor their namespace and not all running applications), which works with a SHA256 token that has to be sent along. By sending an "admin" token, each user could monitor all applications, which we want to limit to dedicated namespaces. 

Our initial idea was to implement a custom authentication method for prometheus and loki datasources. In order to achieve this out solution so far is a TokenExchange via a keycloak that communicates with the Openshift to exchange the keycloak bearer token with an openshift api-token in SHA256 format. For testing purposes we have written a script that exchanges the keycloak for the openshift internal token. This approach worked and using this token as an authorization header in grafana when communicating with the datasources had the intended effect: we can access the datasources and only see metrics/logs for namespaces we are authorized to. 

However, the problem we currently face is that we don't know how to introduce this TokenExchange to Grafana. 
A first approach was to rewrite the ForwardOAuthIdentity and implement a TokenExchange, but this is not used for reasons I can't explain (pkg/infra/httpclient/httpclientprovider/forwarded_oauth_identity_middleware.go).

From our research so far, we have identified multiple approaches that might or might not work due to our limited understanding of grafanas internal workings and plugin development for grafana: 

* Writing a custom datasource plugin for prometheus and loki which handle the token exchange (not sure whether we need a backend or datasource plugin, both?)
* Creating a plugin, that wraps calls to loki and prometheus datasources which performs the token exchange before calling the default datasources for queriying data. 
* something else completely? 

Any pointers at which direction would be the way to go and maybe some pointers/resources at implementation hints would be greatly appreciated.

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
